CI: Add all-in-one macOS build script

PatTheMav opened the pull request.

DDRBoxman commented on CI/full-build-macos.sh:
```diff
@@ -0,0 +1,453 @@
+        -x ./OBS.app/Contents/PlugIns/text-freetype2.so \
+        -x ./OBS.app/Contents/PlugIns/obs-libfdk.so \
+        -x ./OBS.app/Contents/PlugIns/obs-outputs.so
+
```
Looks like you need to still copy over qtnetwork
https://github.com/obsproject/obs-studio/blob/master/CI/install/osx/packageApp.sh#L53

PatTheMav replied:
I always wondered why the QtNetwork framework was first removed, then added again?

Apparently the updated dylibBundler copied the QtNetwork framework into the bundle automatically, there was no need to add it again (except for the reason why the current scripts manually remove and add it).

DDRBoxman commented on CI/full-build-macos.sh:
```diff
@@ -0,0 +1,453 @@
+    install_name_tool -change libmbedcrypto.3.dylib @executable_path/../Frameworks/libmbedcrypto.3.dylib ./OBS.app/Contents/Plugins/obs-outputs.so
+    install_name_tool -change libmbedx509.0.dylib @executable_path/../Frameworks/libmbedx509.0.dylib ./OBS.app/Contents/Plugins/obs-outputs.so
+
+    ${CI_SCRIPTS}/app/dylibBundler_v2 -cd -of -a ./OBS.app -q -f \
```
is dylibBundler_v2 just using the latest code from the master branch of dylibBundler?

PatTheMav replied:
Not exactly - it's a fork that went ahead and rewrote big parts of the code, so much so that the original dev made that person the new maintainer. I compiled the new binary from the most current branch: https://github.com/SCG82/macdylibbundler

kenlowrie commented:
Okay, first look, love this. Freaking awesome. I've still got a number of scenarios I want to test, including the codesigning, but my initial testing quite the success. Worked perfecting with the full build, even running multiple times didn't trip it up.

Enhancement Request 1: Add a switch that allows me to bypass the dependency download/setup/compile step. -s is great for moving straight to the bundle create or packaging step, but while actively working on something, don't want to keep download the deps over and over.

-sb worked great, multiple times, even if I held the current OBS.app open, it still built a new one.

-sp worked, but side effect it moved OBS.app from the build directory to the root of the source tree. So running -sp again fails, because the bundle is no longer there. Quick code review wasn't obvious why that side effect happened, but will investigate further.

Probably should add CI/scripts/Brewfile.lock.json to .gitignore

Okay, that's it for now, will continue working with it, but this thing rocks! Thank you for creating it!

PatTheMav commented:
@kenlowrie Thanks for the feedback mate. 

As for the dependency download skipping, the `curl -L` switch should handle it, as it "resumes" the downloads. If the archives are still in the `obs-build-dependencies` folder, it will display downloads, but they should more or less skip any downloading once the file size on disk is detected to be identical to the file size reported by the server.

As for the `-sp` switch, I understand that the desired outcome to have a signed copy of the built app bundle instead of the artifact being packaged to a `.dmg` and moved away (the move happens on line 257 as the `dmgbuild` expects it there).

kenlowrie commented on CI/full-build-macos.sh:
```diff
@@ -0,0 +1,453 @@
+        exit 1
+    fi
+
+    mv ./build/OBS.app ./OBS.app
```
Why move the bundle to the root here? Shouldn't it stay in the build folder?

PatTheMav commented:
Yeah I'll probably copy it instead and clean up orphaned packaging artifacts after the `.dmg` is created. Good feedback.

kenlowrie commented:
> Yeah I'll probably copy it instead and clean up orphaned packaging artifacts after the `.dmg` is created. Good feedback.

Okay, cool. One more side effect I just noticed, when I do the build again after doing the -p, because the bundle isn't there, it recompiles a bunch of stuff that it normally doesn't do... So it will save time by doing that. Thanks for your consideration!

kenlowrie commented:
> @kenlowrie Thanks for the feedback mate.
> 
> As for the dependency download skipping, the `curl -C -` switch should handle it, as it "resumes" the downloads. If the archives are still in the `obs-build-dependencies` folder, it will display downloads, but they should more or less skip any downloading once the file size on disk is detected to be identical to the file size reported by the server.

Okay, gotcha on the re-downloading of the files, that definitely works exactly like you said. The attempted install of the brew components and the remake of the CEF framework (although it doesn't really do anything), still does consume extra cycles. As a quick test, I added -x, which sets SKIP_DEPS, and then I changed full-build-macos() as:

full-build-macos() {
    if [ -z "${SKIP_DEPS}" ]; then
        hr "Installing Homebrew dependencies"
        install_homebrew_deps

        for DEPENDENCY in "${BUILD_DEPS[@]}"; do
            set -- ${DEPENDENCY}
            trap "caught_error ${DEPENDENCY}" ERR
            FUNC_NAME="install_${1}"
            ${FUNC_NAME} ${2}
        done
    fi

    check_ccache
    configure_obs_build
    run_obs_build

Without -x, a build takes 30 seconds on my machine. With -x, it takes about 4 seconds. Not a huge amount of time, but probably would add up after a dozen or more compiles...

Just wanted to explain my reasoning a little more, but I respect your decision either way. I will move on to testing the codesigning now, after I go get another cup of coffee...

kenlowrie commented:
I tried to run -bc to test codesigning, and it prompted me for id/pw/identity (which hopefully I answered correctly), it proceeded to replace the existing signature in sparkle and OBS.app, but then it failed with the following messages:

`Stored the password in the keychain item 'OBS-Codesign-Password'.
./OBS.app/Contents/Frameworks/Sparkle.framework/Versions/A/Resources/Autoupdate.app/Contents/MacOS/fileop: replacing existing signature
./OBS.app/Contents/Frameworks/Sparkle.framework/Versions/A/Resources/Autoupdate.app/Contents/MacOS/Autoupdate: replacing existing signature
./OBS.app/Contents/Frameworks/Sparkle.framework: replacing existing signature
./OBS.app: edited signature app bundle with Mach-O thin (x86_64) [com.obsproject.obs-studio]
2020-04-25 10:45:52.194 altool[95291:4020973] *** Error: Unable to notarize app.
2020-04-25 10:45:52.194 altool[95291:4020973] *** Error: code -22016 (Unable to validate your application. We are unable to create an authentication session.)
───────────────────────────────────────────────────────────────────────
[OBS-Studio] 	ERROR: Upload of application bundle for notarization failed
───────────────────────────────────────────────────────────────────────`
I likely did something wrong, have tried to codesign anything for a few years (since the last time I uploaded an iOS program to the app store), and I did that signing inside Xcode, not via the command line...

For now, I'm going to rm -rf the sparkle framework, and then rebuild everything without codesigning, to make sure I'm still good. If you have any suggestions or things you want me to try with codesigning or anything else, just let me know.

PatTheMav commented:
Did you use an application-specific password or your actual Apple ID password? For Notarization you'd need to create a new application-specific password, as there is no 2FA for CLI tools.

kenlowrie commented:
> Did you use an application-specific password or your actual Apple ID password? For Notarization you'd need to create a new application-specific password, as there is no 2FA for CLI tools.

That was the first problem, which I stumbled upon and fixed, and during the second go round, after a few minutes it failed because I needed to login and accept some new terms and conditions, which I just did, and I'm about to run it again. :)

kenlowrie commented:
`The staple and validate action worked!
./OBS.app: accepted
source=Notarized Developer ID
───────────────────────────────────────────────────────────────────────
[OBS-Studio] 	Skipping Apple DMG creation
───────────────────────────────────────────────────────────────────────`
Looks like it worked just fine!

kenlowrie commented:
Not sure if it's possible or makes sense, but if there's a way to verify whether or not the "OBS-Codesign-Password" is already in the keychain and just use it, rather than always prompting me to enter the same info, that would be a nice-to-have. May need a switch to tell the system to look and use it if exists, since the default behavior of prompting is probably the better default action.

In any case, I doubt I would do this very often anyway, so it's more of a matter for the person that would be doing this a lot, and if said person doesn't exist, then it's probably a waste of time.

Thanks again for allowing me to test this new build script. I will keep watch for any script updates and retest, otherwise I'm looking forward to this being pulled into master on the main repo. This is going to save anyone that builds the macos version a TON of time and frustration trying to figure it out the first time... Great job!

PatTheMav commented:
@kenlowrie In theory it should be possible to fetch the entry via the `security` command, but I hesitate to implement it.

Code-signing and notarization are only really necessary for distribution and I'd like that step to be a very deliberate choice, _especially_ as it requires the script to ask for account information. As such I was already investigating whether I could remove the keychain item after a successful run to leave the system in the state it was at when the script was invoked.

As for the dependencies I think that those steps could be skipped if building is set to be skipped as well (which would turn the script into a bundling/packaging helper in that case), with an additional check for a pre-existing CEF build. 

Any other customizations seem out-of-scope to me, as there are too many options to consider otherwise. But in general once all default dependencies are available, any different combination can be compiled and run by changing the `cmake` call, followed by `make -j4`. The wiki could be updated with some common examples for those CMake configuration though. I'll look into that some time.

PatTheMav commented:
@kenlowrie I moved quite a lot of things around, simplified output, updated notarization, etc.

If you have some time to spare, it'd be great if you could run it for yourself and let me know your thoughts.

Things to consider:

* The script will probably not facilitate quick "change&rebuild" functionality, I reckon running `cmake [..] && make -j4` in the build directory should be quicker and easier anyway
* Notarization uses `xcnotary` now, which should be a bit quicker and easier to use
* I've simplified the prompts and moved explanations for notarization requirements into doc comments
* Packaging and code-signing implies a re-built, as does notarization as the script needs to ensure existing code signatures
* If no packaging is chosen, only the app bundle is notarized; otherwise the `.dmg` image is used for notarization as according to Apple everything should be code-signed from the inside out, then the outer-most container needs to be notarized and stapled (`xcnotary` takes care of that)

PatTheMav commented:
(I'm also aware that the branch contains commits from another PR, I will re-base once https://github.com/obsproject/obs-studio/pull/2877 is merged upstream).

kenlowrie commented:
> (I'm also aware that the branch contains commits from another PR, I will re-base once [obsproject#2877](https://github.com/obsproject/obs-studio/pull/2877) is merged upstream).

Okay, I have been out of the office all day, but as soon as I get back, I will take a look at it. If possible, this evening, otherwise I'll get on it first thing tomorrow.

I'll let you know how it goes.

kenlowrie commented:
Have some additional notes on all tests run thus far if interested, although everything except notarization is running as expected. Here's the rundown on what I've seen with the notarization:

`CI/full-build-macos.sh -sn`

**1st Attempt:**

If you provide invalid input on "Apple developer identity" (which I did accidentally) - fails signing sparkle, cef, obs with no identify found, but script continues on to request account id (Notarization setup). 

```
[OBS-Studio] Code-signing application bundle
[OBS-Studio] Code-signing Setup
  + Apple developer identity: "Developer ID Application: Kenneth Lowrie (XXXXXXXXXX)"
  + Code-sign Sparkle framework...
"Developer ID Application: Kenneth Lowrie (XXXXXXXXXX)": no identity found
"Developer ID Application: Kenneth Lowrie (XXXXXXXXXX)": no identity found
"Developer ID Application: Kenneth Lowrie (XXXXXXXXXX)": no identity found
  + Code-sign CEF framework...
"Developer ID Application: Kenneth Lowrie (XXXXXXXXXX)": no identity found
"Developer ID Application: Kenneth Lowrie (XXXXXXXXXX)": no identity found
"Developer ID Application: Kenneth Lowrie (XXXXXXXXXX)": no identity found
"Developer ID Application: Kenneth Lowrie (XXXXXXXXXX)": no identity found
"Developer ID Application: Kenneth Lowrie (XXXXXXXXXX)": no identity found
  + Code-sign OBS code...
"Developer ID Application: Kenneth Lowrie (XXXXXXXXXX)": no identity found
CI/full-build-macos.sh: line 401: 25546 Segmentation fault: 11  codesign -v -verbose=4 ./OBS.app
[OBS-Studio] Skipping Application bundle creation
[OBS-Studio] Skipping Apple DMG creation
[OBS-Studio] Notarizing OBS for macOS
[OBS-Studio] Notarization Setup
  + Apple account id: ^C

```

**_Expected behavior**: It should fail (stop) upon first error._

**2nd attempt (with valid dev ident):**

Success, until "Run xcnotary", which fails with "not found".

```
[OBS-Studio] Code-signing application bundle
[OBS-Studio] Code-signing Setup
  + Apple developer identity: Developer ID Application: Kenneth Lowrie (XXXXXXXXXX)
  + Code-sign Sparkle framework...
./OBS.app/Contents/Frameworks/Sparkle.framework/Versions/A/Resources/Autoupdate.app/Contents/MacOS/fileop: replacing existing signature
./OBS.app/Contents/Frameworks/Sparkle.framework/Versions/A/Resources/Autoupdate.app/Contents/MacOS/Autoupdate: replacing existing signature
./OBS.app/Contents/Frameworks/Sparkle.framework: replacing existing signature
  + Code-sign CEF framework...
  + Code-sign OBS code...
./OBS.app: edited signature app bundle with Mach-O thin (x86_64) [com.obsproject.obs-studio]
[OBS-Studio] Skipping Apple DMG creation
[OBS-Studio] Notarizing OBS for macOS
[OBS-Studio] Notarization Setup
  + Apple account id: k****e@gmail.com
  + Apple developer password: 
The keychain item 'OBS-Codesign-Password' already has the same password. No changes were made.
  + Run xcnotary...
CI/full-build-macos.sh: line 508: xcnotary: command not found
[OBS-Studio] Skipping macOS notarization

```

xcnotary not found.

Expected behavior: I think that the homebrew setup should have accounted for this. Or maybe something else, not sure. For now, I worked around with:

`brew install akeru-inc/tap/xcnotary`

**3rd attempt**

xcnotary aborts. output:

```
[OBS-Studio] Notarizing OBS for macOS
[OBS-Studio] Notarization Setup
  + Apple account id: k****e@gmail.com
  + Apple developer password: 
The keychain item 'OBS-Codesign-Password' already has the same password. No changes were made.
  + Run xcnotary...
Processing...
 
✔ Perform check: Developer ID signing
✔ Perform check: Hardened runtime
✔ Perform check: No get-task-allow entitlement
✔ Perform check: Secure timestamp

thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 2, kind: NotFound, message: "No such file or directory" }', src/notarize/run.rs:116:22
[OBS-Studio] Skipping macOS notarization
ken:/Users/Shared/data/src/obs/PatTheMav/2/obs-studio $ 

```
Expected behavior: 

1. should complete, but this may be due to my xcnotary version, setup, config ... 
2. shouldn't it stop on that error, instead of continuing with "Skipping macOS notarization"?  

Not really sure, so just let me know.

PatTheMav commented:
Thanks for the test, helps a lot.

As for 1st attempt you're right, usually the whole script should abort upon non-zero return codes, so I'm surprised that it continued - need to check that. The uncaught segfault is especially hurtful.. 😅

As for the 2nd attempt, that's my bad - xcnotary was originally part of the Brewfile but I removed it temporarily as it didn't have `.dmg` support. It should be installed automatically.

Number 3 is also weird (similar to number 1) - a non-zero return code should be caught by the `trap` and quit the script.

As for the other comments, feel free to post them, as this is just my personal fork and the final CR will go to upstream without the comment feed anyway.

PatTheMav commented:
As it turns out, chaining commands in the form of

`[CONDITION] && COMMAND_IF_TRUE || COMMAND_IF_FALSE`

are not covered by `set -e` (as explained by the bash documentation). By moving the checks into the functions, any errors during the steps should now abort the process as expected.

Regarding Xcnotary that's obviously not something the script can fix, but it looks like it failed while trying to run the `ditto` command to copy all files for a zip archive, which is weird.

I pushed some changes, feel free to have another look.

kenlowrie commented:
Okay, the new version solves the invalid input issue. Haven't tested a clean build yet, to see if it handles the brew install of xcnotary, but will get to that.

Initial attempt at notarization with "CI/full-build-macos -sn" failed because it attempted to sign the package, when no package was built. e.g.:

```
  + Check code-sign result...
Executable=/Users/Shared/data/src/obs/PatTheMav/2/obs-studio/build/OBS.app/Contents/MacOS/obs
Identifier=com.obsproject.obs-studio
Format=app bundle with Mach-O thin (x86_64)
CodeDirectory v=20500 size=27109 flags=0x10000(runtime) hashes=838+5 location=embedded
Signature size=9058
Authority=Developer ID Application: Kenneth Lowrie (XXXXXXXXXX)
Authority=Developer ID Certification Authority
Authority=Apple Root CA
Timestamp=May 7, 2020 at 11:19:56 AM
Info.plist entries=16
TeamIdentifier=XXXXXXXXXX
Runtime Version=10.15.4
Sealed Resources version=2 rules=13 files=1213
Internal requirements count=1 size=188
  + Skipping installer image creation
  + No OBS disk image found

```
I temporarily worked around by adding this 2nd test in codesign_image():

```
codesign_image() {
    if [ ! -n "${CODESIGN_OBS}" ]; then step "Skipping installer image code signing"; return; fi
    if [ ! -n "${PACKAGE_OBS}" ]; then step "Skipping installer image code signing because package was not created"; return; fi

    ensure_dir "${CHECKOUT_DIR}/build"

```
And that got me back to the same failure on xcnotary, which I will start debugging now, to see if I can figure out why ditto is failing.

PatTheMav commented:
@kenlowrie Yes that's my bad, the script is supposed to "skip" over bundles or packages if they're not found and try to code-sign and notarize what's available. I changed the checks in the outer functions to return in the case of an error and not fail (whereas the "inner" functions still will fail on unexpected states).

I can confirm the xcnotary issue with just an app bundle and created an issue at the `xcnotary repo: https://github.com/akeru-inc/xcnotary/issues/6

PatTheMav commented:
Found the issue, can fix it in the script. Local paths need to be supplied as `./MY_APP.app` and not `MY_APP.app` for the compression step to work.. ¯\_(ツ)_/¯

kenlowrie commented:
> Found the issue, can fix it in the script. Local paths need to be supplied as `./MY_APP.app` and not `MY_APP.app` for the compression step to work.. ¯_(ツ)_/¯

Cool. Funny thing, I spent this afternoon setting up my dev system to be able to build and debug Rust apps, and had just now found the solution, and was coming over here to let you know, only to find out that you already fixed it like 3 hours ago. LoL.

I will update my local repo and do some more testing, starting with a clean build and no xcnotary installed. Probably wont' finish up until tomorrow, though, as I will be OOO in the morning.

Thanks,
-ken

kenlowrie commented:
Haven't tried to see why this is failing, just kicked off a new build from a fresh clone of the macos-build-scripts branch...

```
ken:/Users/Shared/data/src/obs/PatTheMav/2/obs-studio $ CI/full-build-macos.sh 
[OBS-Studio] Installing Homebrew dependencies
Unlinking /usr/local/Cellar/swig/3.0.12... 3 symlinks removed
Unlinking /usr/local/Cellar/qt/5.14.1... 497 symlinks removed
Using akeru-inc/tap
Using jack
Using speexdsp
Using cmake
Using freetype
Using fdk-aac
######################################################################## 100.0%
Installing https://gist.githubusercontent.com/ddrboxman/9c7a2b08933166f4b61ed9a44b242609/raw/ef4de6c587c6bd7f50210eccd5bd51ff08e6de13/qt.rb
######################################################################## 100.0%
Installing https://gist.githubusercontent.com/ddrboxman/4cada55c51803a2f963fa40ce55c9d3e/raw/572c67e908bfbc1bcb8c476ea77ea3935133f5b5/swig.rb
Installing akeru-inc/tap/xcnotary
Homebrew Bundle complete! 9 Brewfile dependencies now installed.
[OBS-Studio] Setting up pre-built macOS OBS dependencies v2020-04-24
  + Download...
  + Unpack...
tar: Error opening archive: Unrecognized archive format
  + ERROR during build step: obs-deps
ken:/Users/Shared/data/src/obs/PatTheMav/2/obs-studio $ 
```

kenlowrie commented:
```
<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>SignatureDoesNotMatch</Code><Message>The request signature we calculated does not match the signature you provided. Check your key and signing method.</Message><AWSAccessKeyId>AKIAIWNJYAX4CSVEH53A</AWSAccessKeyId><StringToSign>AWS4-HMAC-SHA256
20200507T223256Z
20200507/us-east-1/s3/aws4_request
67df8e41af81536aa48bf6ac82a976a97b667b9be162f4ef0bfc6488bea60b05</StringToSign><SignatureProvided>f3ad316eb2165d3cffe2b0174b0b4b7d2ccd4b4c4c70e29dfc15bbf5ae9b1de4</SignatureProvided><StringToSignBytes>41 57 53 34 2d 48 4d 41 43 2d 53 48 41 32 35 36 0a 32 30 32 30 30 35 30 37 54 32 32 33 32 35 36 5a 0a 32 30 32 30 30 35 30 37 2f 75 73 2d 65 61 73 74 2d 31 2f 73 33 2f 61 77 73 34 5f 72 65 71 75 65 73 74 0a 36 37 64 66 38 65 34 31 61 66 38 31 35 33 36 61 61 34 38 62 66 36 61 63 38 32 61 39 37 36 61 39 37 62 36 36 37 62 39 62 65 31 36 32 66 34 65 66 30 62 66 63 36 34 38 38 62 65 61 36 30 62 30 35</StringToSignBytes><CanonicalRequest>GET
/196753436/71a77500-86f9-11ea-8ffb-29a643173163
X-Amz-Algorithm=AWS4-HMAC-SHA256&amp;X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20200507%2Fus-east-1%2Fs3%2Faws4_request&amp;X-Amz-Date=20200507T223256Z&amp;X-Amz-Expires=300&amp;X-Amz-SignedHeaders=host&amp;actor_id=0&amp;repo_id=196753436&amp;response-content-disposition=attachment%3B%20filename%3Dosx-deps-2020-04-24.tar.gz&amp;response-content-type=application%2Foctet-stream
host:github-production-release-asset-2e65be.s3.amazonaws.com

host
UNSIGNED-PAYLOAD</CanonicalRequest><CanonicalRequestBytes>47 45 54 0a 2f 31 39 36 37 35 33 34 33 36 2f 37 31 61 37 37 35 30 30 2d 38 36 66 39 2d 31 31 65 61 2d 38 66 66 62 2d 32 39 61 36 34 33 31 37 33 31 36 33 0a 58 2d 41 6d 7a 2d 41 6c 67 6f 72 69 74 68 6d 3d 41 57 53 34 2d 48 4d 41 43 2d 53 48 41 32 35 36 26 58 2d 41 6d 7a 2d 43 72 65 64 65 6e 74 69 61 6c 3d 41 4b 49 41 49 57 4e 4a 59 41 58 34 43 53 56 45 48 35 33 41 25 32 46 32 30 32 30 30 35 30 37 25 32 46 75 73 2d 65 61 73 74 2d 31 25 32 46 73 33 25 32 46 61 77 73 34 5f 72 65 71 75 65 73 74 26 58 2d 41 6d 7a 2d 44 61 74 65 3d 32 30 32 30 30 35 30 37 54 32 32 33 32 35 36 5a 26 58 2d 41 6d 7a 2d 45 78 70 69 72 65 73 3d 33 30 30 26 58 2d 41 6d 7a 2d 53 69 67 6e 65 64 48 65 61 64 65 72 73 3d 68 6f 73 74 26 61 63 74 6f 72 5f 69 64 3d 30 26 72 65 70 6f 5f 69 64 3d 31 39 36 37 35 33 34 33 36 26 72 65 73 70 6f 6e 73 65 2d 63 6f 6e 74 65 6e 74 2d 64 69 73 70 6f 73 69 74 69 6f 6e 3d 61 74 74 61 63 68 6d 65 6e 74 25 33 42 25 32 30 66 69 6c 65 6e 61 6d 65 25 33 44 6f 73 78 2d 64 65 70 73 2d 32 30 32 30 2d 30 34 2d 32 34 2e 74 61 72 2e 67 7a 26 72 65 73 70 6f 6e 73 65 2d 63 6f 6e 74 65 6e 74 2d 74 79 70 65 3d 61 70 70 6c 69 63 61 74 69 6f 6e 25 32 46 6f 63 74 65 74 2d 73 74 72 65 61 6d 0a 68 6f 73 74 3a 67 69 74 68 75 62 2d 70 72 6f 64 75 63 74 69 6f 6e 2d 72 65 6c 65 61 73 65 2d 61 73 73 65 74 2d 32 65 36 35 62 65 2e 73 33 2e 61 6d 61 7a 6f 6e 61 77 73 2e 63 6f 6d 0a 0a 68 6f 73 74 0a 55 4e 53 49 47 4e 45 44 2d 50 41 59 4c 4f 41 44</CanonicalRequestBytes><RequestId>992A3704D2EE63A8</RequestId><HostId>VMBz9h7HMu/qj3PanKCEPJ7v5BR1IvkPIVho73C9tR712itoJoHok6WtefcgA+wwoxZts1e/mPE=</HostId></Error>
```

PatTheMav commented:
Annoying case of "works for me", maybe Amazon had a little hiccup there.

kenlowrie commented:
Figures, but not surprising. I will try again when I get into my office later today. Sorry for the false alarm.

kenlowrie commented:
Okay, everything seems to be working as expected now. The prior issue was some sort of Amazon glitch, because its worked several times today with no issues.

My only remaining question at this point is what the expectations are when notarizing. The way I interpreted the docs in the shell script, if notarization is requested, then the bundle will be notarized (as it is when running with **_-bn_**). However, if the -p switch is also specified, then my interpretation was that both the bundle and the image would be notarized. This is not what I am seeing. If I pass **_-sbpn_**, everything is code signed as expected, but only the image is notarized.

Further, if I have previously built the image, but then invoke the script with **_-sbn_**, it will once again notarize the image, even though I didn't rebuild it during this step. Seems like it should have instead notarized the bundle, as would be the case if no image existed.

With all that said, I'm not so sure that this is anything to worry about. Seems like practically speaking, it will accomplish what it needs to for the main usage cases. One exception might be to the 1st point above, if, in fact, both the bundle and the image should have been notarized when the "**_-bpn_**" switches are all specified.

For me, this thing works perfectly for what I need as-is. So whatever you decide needs to happen (if anything), is fine. :)

Other comments:

1. I really like the different colored messages for the main sections and the individual steps. It makes it so much easier to follow along with that is going on, and quickly scroll through the output when I'm looking for something.
2. It might b nice if you told me specifically what was being notarized in the notarize_macos() function. Now it just says "Notarizing OBS for macOS", whether it is notarizing the bundle or the image. If maybe you just changed "Run xcnotary..." to "Run xcnotary ${NOTARIZE_TARGET}..." this would be helpful to me. But purely a nice to have, not a big deal either way.

Appreciate you letting me help vet this new build system. Cant wait till it's pulled into the master upstream. It will solve so many frustrations on building from source, especially for newbies, but also for those familiar with the process.

If you have any questions or need any further testing done, please let me know.

PatTheMav commented:
Thanks for the feedback, I'll try to explain the current logic:

Right now the script tries to be lazy, meaning it'll "build" by default, but only bundle, package, code-sign, or notarize on-demand.

If given an on-demand option, it will take what is "there" and apply it. So if you want to code-sign and there is a bundle, it gets signed. If there is a package and you want to notarize, it notarizes only the package.

If you ask it to package, but there is no bundle, it will "skip" the step with an error message. If you ask to code-sign, but there is no bundle, it will do the same, but if there is a package, it will try to code-sign it.

As for notarization the current idea is that if a package is there, that will be notarized, as the contained bundle will be notarized as well (so no need to do both). My idea was that people would usually run `-b -p -n` to build, package and notarize (with the latter automatically including code-signing). I removed the implicit build-step triggered by `-p` as I found it more natural to first build and test and package+notarize later (which will code-sign an existing bundle anyway).

So for your example where you want to notarize the bundle _only_ you need to make sure that there is no package there, because an existing `.dmg` will take precedence.

---

Re. the notarization output: Good point, will make it more informative.

I _also_ have to cover the case if `-b` and `-s` are invoked and there is no successful build available to bundle. Will fix both and open a proper PR on upstream.

Thanks again for the reviews and comments.

kenlowrie commented:
Another thing I just thought of you _might_ want to consider.

If -s is not specified, and the bundle and/or package exist, they should be removed?

My logic is, if on a previous pass, I bundled and/or made an image, and then discovered something else I needed to fix, which I then did by simply building, testing, etc., and then I did a -sn without bundling or packaging, then it would end up signing and notarizing a stale OBS.

This is an edge case, and probably not too likely (at least forgetting to bundle), but I could see forgetting to package again, after rebuilding OBS, and end up notarizing a package with the wrong OBS bundle...

I haven't actually tried this yet, it just seems like a possible case from the various test runs I've done. But again, how likely is it? Probably not very...

PatTheMav commented:
The bundle is removed on every invocation of `-b`, the same goes for the package on `-p`, but not a re-build, that is true and it would actually make sense to at least move them away (in a prior version the package was moved to `nightly` in the project root).

I'm now creating a nightly-dir for every successful build and move the available artifacts there, if an actual rebuild is triggered. Testing it right now and will push another (hopefully last) update.. 😄

kenlowrie commented:
Okay, this is odd. I might be caught in some weird state, I cloned the repo fresh from your fork about 20 minutes ago, and I keep hitting this error:

```
[OBS-Studio] Setting up dependency VLC v3.0.8
 + skipping this slow-ass download...
[OBS-Studio] Setting up dependency Sparkle v1.23.0 (might prompt for password)
  + Download...
  + Unpack...
  + Copy to destination...
 + Warning - Sparkle framework already found in /Library/Frameworks
ccache -s >/dev/null 2>&1 && echo "CCache available." || echo "CCache is not available."
 + CCache available.
date +"%Y-%m-%d@%H%M%S"
basename $(find . -name "*.dmg")
find . -name "*.dmg"
usage: basename string [suffix]
       basename [-a] [-s suffix] string [...]
caught_error sparkle 1.23.0
  + ERROR during build step: sparkle
ken:/Users/Shared/data/src/obs/PatTheMav/2/obs-studio $ 
```
So two things. 1) possibly an error with basename (BTW, I did a "set -v" right after sparkle, so I could figure out where it was failing, because 2) it says "sparkle failed, but it's actually in configure_obs_build().

PatTheMav commented:
Yeah if there is no `.dmg` then find will return an empty string, bad oversight. Fixing it right now..

kenlowrie commented:
Okay, ran a whole slew of tests with various combinations. Tried very hard to break it. :) Only found 2 _very minor_ issues that might be considered next time this is updated...

In configure_obs_build(), before checking for ${PACKAGE_NAME}, need to move back to the build directory to handle the case where the bundle existed, and you moved over to the nightly dir... Otherwise, the -f test fails, because we are no longer in the right directory. In that case, we accidentally leave the stale package in place when it should have been moved out.

```
    if [ -d ./OBS.app ]; then
        ensure_dir "${NIGHTLY_DIR}"
        mv ../build/OBS.app .
        info "You can find OBS.app in ${NIGHTLY_DIR}"
    fi
    ensure_dir "${CHECKOUT_DIR}/build"
    if ([ -n "${PACKAGE_NAME}" ] && [ -f ${PACKAGE_NAME} ]); then
        ensure_dir "${NIGHTLY_DIR}"
        mv ../build/$(basename "${PACKAGE_NAME}") .
        info "You can find ${PACKAGE_NAME} in ${NIGHTLY_DIR}"
    fi
```
Like I said, very minor. :)

Second, might need to consider either 1) breaking out codesign as a separate step, or 2) when notarization is specified, forcing the package to be remade. Here's what I'm seeing.

Build, test, etc. i.e. running the script w/no parameters.
Bundle, test i.e. -s -b
Package, test -s -p
Great, everything looks good.

Let's codesign and notarize, -s -n

It proceeds to codesign everything, the bundle, the frameworks, the DMG. It then uploads the .dmg for notarization, but the issue is that the bundle within the package is not the codesigned one... So it fails during notarization due to all the stuff inside the .dmg bundle not being signed.

So the workaround is simply to -s -p -n

Besides that, it works fantastic! Again, I don't think either are showstoppers, better to get the code out there and see what else shakes out. This thing is so much better that what we currently have, I personally would rather not delay any further.

If codesign was a separate flag, and I knew I was headed down the path of notarization, I could instead request the codesigning on the package step (-s -p -c?), so the bundle in the package was properly signed, then after testing, I could do the -s -n.

Hope this is helpful and accurate. Have a great weekend!

PatTheMav commented:
You're right, it needs to `ensure_dir` the build directory before checking for the package, that needs fixing asap.

The code-signing gotcha I'm aware of, which is why I added it to the note at the bottom. I will think about an elegant solution for this and patch it at a later point.

tourdownunder commented:
I found this fork only as I want to build a copy of [obs-mac-virtualcam and cam across [obs-mac-virtualcam-issue-70](https://github.com/johnboiles/obs-mac-virtualcam/issues/70) and they recommend building from source. 

So I'll share this attempt at building your fork in case it is useful

```sh
CI/full-build-macos.sh 
```

Relevant Error seems to be:

```sh
...
[  3%] Building C object deps/libcaption/CMakeFiles/caption.dir/src/caption.c.o
/usr/local/Cellar/swig/4.0.1/share/swig/4.0.1/typemaps/swigtype.swg:608: Error: Syntax error in input(1).
make[2]: *** [deps/obs-scripting/obspython/CMakeFiles/_obspython.dir/obspythonPYTHON.stamp] Error 1
make[2]: *** Deleting file `deps/obs-scripting/obspython/CMakeFiles/_obspython.dir/obspythonPYTHON.stamp'
make[1]: *** [deps/obs-scripting/obspython/CMakeFiles/obspython_swig_compilation.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
...
```

I came across [similar-error-in-google-or-tools-issue](https://github.com/google/or-tools/issues/464). Though don't know where to add the following if required.

```sh
# make python 3.7 the default python 3 and not python 3.8
export PATH="/Library/Frameworks/Python.framework/Versions/3.7/bin:${PATH}"
cmake -DPYTHON_INCLUDE_DIR=/Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m /Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/config-3.7m-darwin .. && make -j
```

PatTheMav commented:
@tourdownunder It seems this might be primarily due to the script not allowing "custom" `cmake` flags.

The script (as it is) fulfills the need to do a default build, so when you want custom flags, you'd still need to run `cmake` and `make` yourself until you're happy with the build, then of course you can use the script to bundle and/or package again.

_Sometimes_  CMAKE accepts flags via environment vars, so

```
PYTHON_INCLUDE_DIR="PATH" ./CI/full-build-macos.sh
```

might work.

kenlowrie commented:
> I found this fork only as I want to build a copy of [obs-mac-virtualcam-issue-70](https://github.com/johnboiles/obs-mac-virtualcam/issues/70) and they recommend building from source. And since I had errors on Mac OS X I searched for other active issues / forks.
> 
> So I'll share this attempt at building your fork in case it is useful. Sorry for the noise if its not useful.
> 
> ```shell
> CI/full-build-macos.sh 
> ```
> 
> Relevant Error seems to be:
> 
> ```shell
> ...
> [  3%] Building C object deps/libcaption/CMakeFiles/caption.dir/src/caption.c.o
> /usr/local/Cellar/swig/4.0.1/share/swig/4.0.1/typemaps/swigtype.swg:608: Error: Syntax error in input(1).
> make[2]: *** [deps/obs-scripting/obspython/CMakeFiles/_obspython.dir/obspythonPYTHON.stamp] Error 1
> make[2]: *** Deleting file `deps/obs-scripting/obspython/CMakeFiles/_obspython.dir/obspythonPYTHON.stamp'
> make[1]: *** [deps/obs-scripting/obspython/CMakeFiles/obspython_swig_compilation.dir/all] Error 2
> make[1]: *** Waiting for unfinished jobs....
> ...

it's odd that swig 4.0.1 was used, that could be the cause of the first error. It should be using 3.0.12, which the first part of the script would have seen to. Can we see the first part of the output where the setup is done? Also, when CMAKE starts running (Header is [OBS-Studio] Run CMAKE for OBS...), shortly after you will see a Found SWIG: ...

What does it show there?

I have forcefully tried to trick the build script into using 4.0.1, but for me, it always figures out and installs/configures the right one...

Anyhow, hope this is helpful.

kenlowrie commented:
Until the PR is merged upstream, is all I need to use this with the current versions of OBS just the full-build-macos.sh script? As far as I can tell the other dependencies are already current in the main repo, but just wanted to ask. Would like to be able to use this script to build with instead of relying on the current ones... Let me know when you have time, thanks!

PatTheMav commented:
If you don't want or need to notarize OBS then you can use only the script.

`xcnotarize` as well as the `entitlements.plist` are only needed for that use case.
